### PR TITLE
Use `queryString` library in `task/getUrl`

### PR DESCRIPTION
**Pull request by cewald**
### Related Issues

Related #5043

### Short Description and Why It's Useful

The new `adjustMultistoreApiUrl` method now uses url-encoding with `queryString.stringifyUrl()`. This leads into the problem that `getUrl()` method can't replace the tokens by simply using `replace`.

I refactored the `getUrl` method to use the `queryString` (`parseUrl`/`stringifyUrl`) library as it already is in us it for the `tokenInHeader` option.

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/lib/sync/task.ts [hunk 1]
```diff
@@ -31,19 +31,25 @@ function _sleep (time) {
...
 function getUrl (task, currentToken, currentCartId) {
-  let url = task.url
-    .replace('{{token}}', (currentToken == null) ? '' : currentToken)
-    .replace('{{cartId}}', (currentCartId == null) ? '' : currentCartId)
+  const parsedUrl = queryString.parseUrl(task.url)
 
-  url = processURLAddress(url); // use relative url paths
-  if (config.storeViews.multistore) {
-    url = adjustMultistoreApiUrl(url)
+  if (parsedUrl.query.token && parsedUrl.query.token === '{{token}}') {
```
Comment (Fifciu) [inline core/lib/sync/task.ts:36]: I think you could use:
```js
if (parsedUrl.query && parsedUrl.query.token === '{{token}}') {
```

You could even go one step forward and use:
```ts
if (parsedUrl?.query?.token === '{{token}}') {
```

File: core/lib/sync/task.ts [hunk 2]
```diff
@@ -31,19 +31,25 @@ function _sleep (time) {
...
+    parsedUrl.query.token = (currentToken == null) ? '' : currentToken
+  }
+
+  if (parsedUrl.query.cartId && parsedUrl.query.cartId === '{{cartId}}') {
```
Comment (Fifciu) [inline core/lib/sync/task.ts:40]: Same as above

**Review by Fifciu — CHANGES_REQUESTED** [review]

File: core/lib/sync/task.ts (hunk 1 above)
Comment (cewald) [reply core/lib/sync/task.ts]: Yes indeed, I've changed it.

**Review by Fifciu — APPROVED** [review]

**Comment by gibkigonzo** [comment]
> This leads into the problem that getUrl() method can't replace the tokens by simply using replace.

What do you mean by that? 🤔
`adjustMultistoreApiUrl` should work in same way as before multistore refactor. As I understood it just replace storeCode in url and should return url. 
Token replacement in `getUrl` takes place before `adjustMultistoreApiUrl` is called so how it is related?

**Comment by cewald** [comment]
`adjustMultistoreApiUrl` is working fine, except it makes `%7B%7Btoken%7D%7D` out of `{{token}}` – and that is why a simple string-replacement, as it was before, won‘t work.

**Comment by gibkigonzo** [comment]
oww ok, I get it. But still `adjustMultistoreApiUrl` is called after `.replace`
replace => https://github.com/DivanteLtd/vue-storefront/blob/develop/core/lib/sync/task.ts#L34-L36
then `adjustMultistoreApiUrl` =>https://github.com/DivanteLtd/vue-storefront/blob/develop/core/lib/sync/task.ts#L40
So when `adjustMultistoreApiUrl` is called then `{{token}}` should be already replaced.

imo your refactor looks better then it was so it's good for me 👍

**Review by gibkigonzo — APPROVED** [review]

**Comment by cewald** [comment]
Just for explanation: `adjustMultistoreApiUrl` is also used in `processLocalizedURLAddress` which again formats the endpoint-urls for the data-resolver-services. That's why it is called first and, if needed, afterwards in the `task/getUrl`.

**Comment by gibkigonzo** [comment]
imo then we need to fix `adjustMultistoreApiUrl`, because `processLocalizedURLAddress` can be used in other places :thinking: Maybe there is some option in `queryString` to not transform `{` into `%7B%7B`

**Comment by cewald** [comment]
Imo there is nothing to fix in `adjustMultistoreApiUrl` – it's actually good and probably more secure to url-encode the query-params using `queryString`. Actually, how it was done before was a bit crude – now it's consistent on both ends.

Anyway, as far as I can see `processLocalizedURLAddress` is currently only used in most data-resolver-services. And `adjustMultistoreApiUrl` only in the `processLocalizedURLAddress` method, the `getFallbackByUrl` action for url-mapping (actually same as data-resolver-service) and the task-lib itself. Should be allright.

Wdyt?

**Comment by Fifciu** [comment]
@cewald please update the changelog and we could merge it

**Comment by cewald** [comment]
@Fifciu done :)

**Comment by Fifciu** [comment]
Great job, thanks!

**Pull request merged**